File: 1010.py
from collections import Counter

class Solution:
    def numPairsDivisibleBy60(self, time: List[int]) -> int:
        count = Counter()
        for t in time:
            count[t % 60] += 1
        
        # print(count)
        total = 0
        total += count[0] * (count[0] - 1) // 2
        for t in range(1, 30):
            total += count[t] * count[60 - t]
        total += count[30] * (count[30] - 1) // 2
        return total
    

File: java/1010.java
/**
 * id: 1010 pairs of songs with total durations divisible by 60
 * tags: array
 * date: 2020-04-22
 * note: Can use an array as the hashmap
 */
class Solution {
    public int numPairsDivisibleBy60(int[] time) {
        Map<Integer, Integer> map = new HashMap<>();
        for(int t : time) {
            int mod = t % 60;
            if(map.containsKey(mod)) {
                map.put(mod, map.get(mod) + 1);
            } else {
                map.put(mod, 1);
            }
        }
        
        int count = 0;
        for(Integer mod : map.keySet()) {
            if(mod == 0 || mod == 30) {
                int c = map.get(mod);
                count += c * (c - 1) / 2;
            } else if (mod < 30) {
                int c1 = map.get(mod);
                int c2 = map.containsKey(60 - mod) ? map.get(60 - mod) : 0;
                count += c1 * c2;
            }
        }
        
        return count;
    }
}
